include/swift/SIL/SILArgumentBuilder.h

```diff
@@ -0,0 +1,120 @@
+//===--- SILArgumentBuilder.h --------------------------------*- c++ -*----===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_SIL_SILARGUMENTBUILDER_H
+#define SWIFT_SIL_SILARGUMENTBUILDER_H
+
+#include "swift/SIL/SILArgument.h"
+#include "swift/SIL/SILBasicBlock.h"
+
+namespace swift {
+
+class SILPhiArgument;
+struct ValueOwnershipKind;
+class ValueDecl;
+class SILType;
+
+class SILArgumentBuilder {
+  SILBasicBlock *block;
+  bool disableEntryBlockVerification;
+
+public:
+  SILArgumentBuilder(SILBasicBlock *block,
+                     bool disableEntryBlockVerification = false)
+      : block(block),
+        disableEntryBlockVerification(disableEntryBlockVerification) {}
+
+  using arg_iterator = SILBasicBlock::arg_iterator;
+
+  // Creation routines for non function arguments.
+#define ARGUMENT_BOILERPLATE(CLASSNAME, PRETTYNAME)                            \
```

Comment:
This API looks specific to phi values. I don't understand why this isn't just a SILPhiArgumentBuilder.

I can think of one pass that replaces function args, AddressLowering, but I don't think it needs to go through the same builder for that special purpose. We will never be replacing or inserting terminator results, so this API is just wrong when we add more argument kinds.

Reply:
Ok, I actually see 2-3 passes that rewrite function signatures using replace/insert. So I don't have a problem centralizing the implementation. I just wish there was a simple SILPhiArgumentBuilder API that didn't have any argument kinds or boilerplate.

And I still don't see how this builder API would make sense for other kinds of block arguments.